src/theme.c

```diff
@@ -56,6 +57,62 @@ drop(struct lab_data_buffer **buffer)
 	}
 }
 
+static cairo_surface_t *
+get_surface(struct lab_data_buffer *buffer)
+{
+	/* Handle CAIRO_FORMAT_ARGB32 buffers */
+	if (buffer->cairo) {
+		return cairo_get_target(buffer->cairo);
+	}
+
+	/* Handle DRM_FORMAT_ARGB8888 buffers */
+	int w = buffer->unscaled_width;
+	int h = buffer->unscaled_height;
+	cairo_surface_t *surface =
+		cairo_image_surface_create(CAIRO_FORMAT_ARGB32, w, h);
+	if (!surface) {
+		return NULL;
+	}
+	unsigned char *data = cairo_image_surface_get_data(surface);
+	cairo_surface_flush(surface);
+	memcpy(data, buffer->data, h * buffer->stride);
+	cairo_surface_mark_dirty(surface);
+	return surface;
+}
+
+static void
+create_hover_fallback(struct theme *theme, struct lab_data_buffer **hover_buffer,
+		struct lab_data_buffer *icon_buffer)
+{
+	assert(icon_buffer);
+	assert(!*hover_buffer);
+
+	cairo_surface_t *icon = get_surface(icon_buffer);
+	int icon_width = cairo_image_surface_get_width(icon);
+	int icon_height = cairo_image_surface_get_height(icon);
+
+	/* FIXME: need to somehow respect rounded corners */
+	int width = SSD_BUTTON_WIDTH;
+	int height = theme->title_height;
+
+	*hover_buffer = buffer_create_cairo(width, height, 1, true);
+
+	cairo_t *cairo = (*hover_buffer)->cairo;
+	cairo_surface_t *surf = cairo_get_target(cairo);
+
+	/* Background */
+	/* FIXME: support downscaling */
+	cairo_set_source_surface(cairo, icon,
+		(width - icon_width) / 2, (height - icon_height) / 2);
+	cairo_paint(cairo);
```

Comment:
I think we should implement downscaling before merging this. Otherwise the hover overlay will become very ugly (and I don't actually know what happens if the `x` and `y` coordinates for cairo become negative).